components/vision.py

```diff
@@ -69,83 +96,151 @@ def reproj(self) -> float:
     def using_multitag(self) -> bool:
         return self.has_multitag
 
+    @feedback
+    def raw_encoder_rotation(self) -> Rotation2d:
+        # The encoder has been set up to return values in the interval [0, 2pi]
+        return Rotation2d(self.encoder.get())
+
+    @feedback
+    def relative_bearing_to_closest_tag(self) -> Rotation2d:
+        # initialise default variables
+        # closest tag, distance, robot position
+        distance = math.inf
+        closest_bearing = Rotation2d()
+
+        for tag in APRILTAGS:
+            robot_to_tag = tag.pose.toPose2d() - self.chassis.get_pose()
+            relative_bearing = robot_to_tag.translation().angle()
+            if (
+                robot_to_tag.translation().norm() < distance
+                and -self.SERVO_HALF_ANGLE
+                <= relative_bearing.radians()
+                <= self.SERVO_HALF_ANGLE
+            ):
+                distance = robot_to_tag.translation().norm()
+                closest_bearing = relative_bearing
+
+        return closest_bearing
+
+    @feedback
+    def desired_turret_rotation(self) -> Rotation2d:
+        # Read encoder angle and account for offset
+        return (
+            self.relative_bearing_to_closest_tag()
+            - self.robot_to_turret.rotation().toRotation2d()
+        )
+
+    def turret_to_servo(self, turret: Rotation2d) -> Rotation2d:
+        return turret - self.servo_offset
+
+    @property
+    def turret_rotation(self) -> Rotation2d:
+        return self.raw_encoder_rotation() - self.encoder_offset
+
+    @property
+    def robot_to_camera(self) -> Transform3d:
+        return Transform3d(
+            self.robot_to_turret.translation(),
+            Rotation3d(
+                self.robot_to_turret.rotation().x,
+                self.robot_to_turret.rotation().y,
+                self.robot_to_turret.rotation().z + self.turret_rotation.radians(),
```

Comment:
nit: avoid creating multiple identical rotation objects
```suggestion
        robot_to_turret_rotation = self.robot_to_turret.rotation()
        return Transform3d(
            self.robot_to_turret.translation(),
            Rotation3d(
                robot_to_turret_rotation.x,
                robot_to_turret_rotation.y,
                robot_to_turret_rotation.z + self.turret_rotation.radians(),
```